internal/armada/server/handlers/submit_jobs_test.go

```diff
@@ -0,0 +1,282 @@
+package handlers
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"testing"
+	"testing/quick"
+
+	"github.com/G-Research/armada/internal/armada/permissions"
+	"github.com/G-Research/armada/internal/armada/repository"
+	"github.com/G-Research/armada/internal/common/auth/permission"
+	"github.com/G-Research/armada/pkg/api"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func TestSubmitJobs(t *testing.T) {
+	properties := map[string]interface{}{
+		"notFound": func(request api.JobSubmitRequest) bool {
+			handler := SubmitJobs(
+				func(_ context.Context, req *api.JobSubmitRequest) ([]*api.Job, error) {
+					return nil, repository.ErrQueueNotFound
+				},
+				func([]*api.Job) (repository.SubmitJobResults, error) {
+					return nil, nil
+				},
+			)
+
+			_, err := handler(context.Background(), &request)
+			if err == nil {
+				t.Errorf("failed to handle an error")
+				return false
+			}
+
+			if status, _ := status.FromError(err); status.Code() != codes.NotFound {
+				t.Errorf("invalid status code: %s", status.Code())
+				return false
+			}
+
+			return true
+		},
+		"internal": func(newJobsError, addJobsError bool, request api.JobSubmitRequest) bool {
+			handler := SubmitJobs(
+				func(_ context.Context, req *api.JobSubmitRequest) ([]*api.Job, error) {
+					if newJobsError {
+						return nil, fmt.Errorf("")
+					}
+					return nil, nil
+				},
+				func([]*api.Job) (repository.SubmitJobResults, error) {
+					if addJobsError {
+						return nil, fmt.Errorf("")
+					}
+					return nil, nil
+				},
+			)
+
+			isError := newJobsError || addJobsError
+
+			_, err := handler(context.Background(), &request)
+			if err == nil && !isError {
+				return true
+			}
+
+			if status, _ := status.FromError(err); status.Code() != codes.Internal {
+				t.Errorf("invalid status code: %s", status.Code())
+				return false
+			}
+
+			return true
+		},
+		"success": func(request api.JobSubmitRequest) bool {
+			handler := SubmitJobs(
+				func(_ context.Context, req *api.JobSubmitRequest) ([]*api.Job, error) {
+					return nil, nil
+				},
+				func(js []*api.Job) (repository.SubmitJobResults, error) {
+					return nil, nil
+				},
+			)
+
+			_, err := handler(context.Background(), &request)
+			if err != nil {
+				t.Errorf("failed to handle request: %s", err)
+			}
+
+			return true
+		},
+	}
+
+	for name, property := range properties {
+		t.Run(name, func(tb *testing.T) {
+			if err := quick.Check(property, nil); err != nil {
```

Comment:
The handler itself is a higher order function (it receives a functions as an input). testing/quick.Check allows generating test cases without manually writing them down. As it is seen from the tests themselves, they are not tied to any input specifically (at least in case of this handler), but are tied to outcome of the injected functions.

with quick package, arbitrary submit requests are generated just to ensure that the higher property of the handler is satisfied no matter how request data looks like. If structure of the request changes in the future, the test would still hold. Quick package ensures that various values for all fields of request are being generated (0 values for numeric types, empty slices and maps, nil values where possible and so on)

Every test here is written for specific outcome (status code) even if there are multiple paths that lead to it.

I would like to see an example that you head in mind though, as I am curious.